### ISSUE TITLE
Use a second branch to maintain Ubuntu compatible script

### ISSUE THREAD

**Issue by LucaZulberti** [eriknellessen/gpg-encrypted-root#4]
Hi I had to modify the script to make it working with my laptop running Ubuntu 18.10.

Look at my [ubuntu-18.10](https://github.com/LucaZulberti/gpg-encrypted-root/tree/ubuntu-18.10) branch.

Are you interested in reviewing it and make it available in your repository as a second branch?

Maybe the script can be rewritten in a way that is compatible for both distros. I have not investigated firther.

**Comment by LucaZulberti** [comment]
Ok, never mind. I hadn't test it enough. It does not work, I will come back when I'll have the solution.

**Comment by eriknellessen** [comment]
Ok, I am excited to see your solution. Do you already know what the difference is between debian stretch and ubuntu 18.10 concerning the mechanisms used by the script?

**Comment by jerabaul29** [comment]
Hi, sorry for digging up, do you know if this somehow works with Ubuntu, or if there is a derived solution that does? :)

**Comment by LucaZulberti** [comment]
Now it is supported already by Ubuntu.

In /etc/crypttab you can add LUKS encrypted partition that are encrypted with a password stored in a GPG encrypted file.

For example:

`sdaX_crypt UUID=<uuid> /etc/keys/cryptkey.gpg luks,keyscript=decrypt_gnupg-sc`

It is recommended to set proper permission to the directory which the key is stored to prevent unwanted delete.

**Comment by jerabaul29** [comment]
Ooh, really nice.

Do you have a link to a modern detailed tutorial that 'does things right', or something in that kind? I have been looking for that all the day but I only can find old articles and blog posts for old Ubuntu and Debian versions. Maybe I am missing something though.

If not such a post exists I can give it a try myself and create a github repo with explanations, but I usually am rather a 'copycat' rather than set up solutions on this kind of things. But can you confirm that the process should be something in this kind? (some links for where to get more infos?):

- set the ```etc/crypttab``` as you show
- have available a private / public keypair, loaded on a smartcard
- encrypt the LUKS password with the public, put it into ```/etc/keys/cryptkey.gpg```
- write the ```decrypt_gnupg-sc``` script so that the LUKS encrypted passwords gets decrypted by the smartcard

or is there something wrong in my understanding?

I think it is not too difficult conceptually, but without documentation, I guess it is easy to get something slightly off here and there ^^ .

**Comment by jerabaul29** [comment]
Ok, searching specifically for ```decrypt_gnupg-sc``` gives me some more interesting hits that the searches I was doing previously today, many thanks, I would not have found this without your help. Will look at this in more details soon, if you know of more places with explanations that will always be welcome:

https://cryptsetup-team.pages.debian.net/cryptsetup/README.gnupg-sc.html

**Comment by LucaZulberti** [comment]
Nice!!! I cannot remember where I found all the necessary information to get the system works.
That link explains correctly the steps I remember I've done:
- setup the encrypted partition
- update crypttab file (the cryptkey.gpg file will be added automatically to intiramfs)
- update initramfs with gpg public key (remember to run update-initramfs utility)

I made everything almost two years ago. If you find any problems, I will try to deepen the topic again. Feel free to share with me anything interesting you will find.
